sched/scheduler/task_scheduler.go

```diff
@@ -13,37 +13,36 @@ type taskAssignments struct {
 // Also returns a modified copy of clusterState.Affinity for the caller to apply (so this remains a pure fn).
 // Does best effort scheduling which tries to assign tasks to nodes already primed for similar tasks.
 // Not all tasks are guaranteed to be scheduled.
-func getTaskAssignments(cs *clusterState, tasks []*taskState) ([]taskAssignments, map[string]*nodeSet) {
-	var tas []taskAssignments
-	affinity := map[string]*nodeSet{}
+func getTaskAssignments(cs *clusterState, tasks []*taskState) ([]taskAssignments, map[string]*nodeGroups) {
+	var assignments []taskAssignments
+	affinity := map[string]*nodeGroups{}
 	snapshotIds := []string{}
 
 	// Create a copy of affinity to modify based on new scheduling.
-	for snapId, set := range cs.affinity {
-		affinity[snapId] = newNodeSet()
-		for nodeId, node := range set.idle {
+	for snapId, groups := range cs.affinity {
+		affinity[snapId] = newNodeGroups()
+		for nodeId, node := range groups.idle {
 			affinity[snapId].idle[nodeId] = node
 		}
-		for nodeId, node := range set.busy {
+		for nodeId, node := range groups.busy {
 			affinity[snapId].busy[nodeId] = node
 		}
 		snapshotIds = append(snapshotIds, snapId)
 	}
 
 	// Loop over all snapshotIDs looking for an idle node. Prefer, in order:
-	// - Hot node for the given snapshotID
-	// - New untouched node (or node run with an empty snapshotID)
-	// - Any node that's hot for a different snapshotID.
-	// Note: tasks are expected to have a valid snapshotID but if it's empty this should still work.
+	// - Hot node for the given snapshotID (one whose last task shared the same snapshotID).
+	// - New untouched node (or node whose last task used an empty snapshotID)
+	// - Any node from the idle pools of nodes associated with other snapshotIDs.
 Loop:
```

Comment:
This is kosher here. People have a knee-jerk reaction to goto's and the like but they have a proper place in clarifying intent for deeply nested loops.